### ISSUE TITLE
Support to use the outer side of the (Anti) Semi Join to build the hash table

### ISSUE THREAD

**Issue by winoros** [pingcap/tidb#32335]
## Enhancement

Currently, when TiDB meets a(an) (Anti) Semi join, it would always choose the inner side to build the hash table.
When the inner side is extremely large compared with the outer side, our current strategy would cost a lot on building the hash table by the inner side.

So we should support constructing the hash table by the outer side to speed up the execution when the data distribution is like the above case.

It's something the same with supporting constructing hash table by the outer side of the outer join, which is implemented in TiDB.

**Comment by winoros** [comment]
We may need to pay attention to the mentioned case above that the outer is very small, and the inner is extremely large.

Suppose that the outer side is just one row. The inner side is 1 billion rows. If we don't set an early break when we find that the outer side is fully matched. We may still need to read many rows from the inner side.

**Comment by fixdb** [comment]
So basically we need to support the following operators in planner/executor:
- left / right semi join
- left / right anti join
- left / right null aware anti join

**Comment by likzn** [comment]
Hi, can i have a try this issue? 
And what we should do is add extra plan like `joins = append(joins, p.getHashJoin(prop, 1, true))`?
Because i ran below test, and enforce to use inner or outer to build hash table(make `getHashJoin` third arg true or false).
```
func TestJoin(t *testing.T) {
	mock, err := mock.NewCluster()
	require.NoError(t, err)
	require.NoError(t, mock.Start())
	defer mock.Stop()

	tk := testkit.NewTestKit(t, mock.Storage)
	tk.MustExec("use test")
	tk.MustExec("create table t1(id int,name char(128));")
	tk.MustExec("create table t2(name char(128));")
	for i := 0; i < 100000; i++ {
		id := strconv.FormatInt(int64(i), 10)
		tk.MustExec("insert into t1(id,name) values(" + id + "," +
			"'" + strings.Repeat("xx", 10) + id + "');")
	}
	tk.MustExec("insert into t2(name) values('name1')")
	tk.MustExec("analyze table t1")
	tk.MustExec("analyze table t2")
	tk.MustQuery("explain analyze SELECT * FROM t2 where exists (SELECT t2.name FROM t1 where t2.name = t1.name)").Check(nil)
}
```
The `build_hash_table` cost gap is big. It seem like that the `executor` is already implemented to chose inner or outer to build table. 
```
use outer to build
// "[HashJoin_9 0.80 0 root  time:54.9ms, loops:1, **build_hash_table:{total:54.8ms, fetch:36.2ms, build:18.7ms}**, probe:{concurrency:5, total:274.3ms, max:54.9ms, probe:12.6µs, fetch:274.3ms} semi join, equal:[eq(test.t2.name, test.t1.name)] 9.78 MB 0 Bytes]
//[├─TableReader_15(Build) 100000.00 100000 root  time:39.1ms, loops:99, cop_task: {num: 1, max: 35.8ms, proc_keys: 0, tot_proc: 35ms, rpc_num: 1, rpc_time: 35.8ms, copr_cache_hit_ratio: 0.00} data:Selection_14 2.58 MB N/A]
//[│ └─Selection_14 100000.00 100000 cop[tikv]  tikv_task:{time:35.1ms, loops:0} not(isnull(test.t1.name)) N/A N/A]
//[│   └─TableFullScan_13 100000.00 100000 cop[tikv] table:t1 tikv_task:{time:35.1ms, loops:0} keep order:false N/A N/A]
//[└─TableReader_12(Probe) 1.00 1 root  time:45.4µs, loops:2, cop_task: {num: 1, max: 116.4µs, proc_keys: 0, rpc_num: 1, rpc_time: 107.6µs, copr_cache_hit_ratio: 0.00} data:Selection_11 174 Bytes N/A]
//[└─Selection_11 1.00 1 cop[tikv]  tikv_task:{time
//

use inner to build
//[HashJoin_9 0.80 0 root  time:55.5ms, loops:1, **build_hash_table:{total:198.1µs, fetch:190.9µs, build:7.17µs}**, probe:{concurrency:5, total:276.8ms, max:55.4ms, probe:17.8ms, fetch:258.9ms} semi join, equal:[eq(test.t2.name, test.t1.name)] 17.5 KB 0 Bytes]
//[├─TableReader_12(Build) 1.00 1 root  time:126.1µs, loops:2, cop_task: {num: 1, max: 234.9µs, proc_keys: 0, rpc_num: 1, rpc_time: 214.8µs, copr_cache_hit_ratio: 0.00} data:Selection_11 174 Bytes N/A]
//[│ └─Selection_11 1.00 1 cop[tikv]  tikv_task:{time:201.2µs, loops:0} not(isnull(test.t2.name)) N/A N/A]
//[│   └─TableFullScan_10 1.00 1 cop[tikv] table:t2 tikv_task:{time:201.2µs, loops:0} keep order:false N/A N/A]
//[└─TableReader_15(Probe) 100000.00 100000 root  time:53.8ms, loops:99, cop_task: {num: 1, max: 49.5ms, proc_keys: 0, tot_proc: 48ms, rpc_num: 1, rpc_time: 49.4ms, copr_cache_hit_ratio: 0.00} data:Selection_14 2.58 MB N/A]
//[└─Selection_14 100000.00 100000 cop[tikv]  tikv_task:{time:
```

**Comment by winoros** [comment]
> The `build_hash_table` cost gap is big. It seem like that the `executor` is already implemented to chose inner or outer to build table.
> 
> ```
> use outer to build
> // "[HashJoin_9 0.80 0 root  time:54.9ms, loops:1, **build_hash_table:{total:54.8ms, fetch:36.2ms, build:18.7ms}**, probe:{concurrency:5, total:274.3ms, max:54.9ms, probe:12.6µs, fetch:274.3ms} semi join, equal:[eq(test.t2.name, test.t1.name)] 9.78 MB 0 Bytes]
> //[├─TableReader_15(Build) 100000.00 100000 root  time:39.1ms, loops:99, cop_task: {num: 1, max: 35.8ms, proc_keys: 0, tot_proc: 35ms, rpc_num: 1, rpc_time: 35.8ms, copr_cache_hit_ratio: 0.00} data:Selection_14 2.58 MB N/A]
> //[│ └─Selection_14 100000.00 100000 cop[tikv]  tikv_task:{time:35.1ms, loops:0} not(isnull(test.t1.name)) N/A N/A]
> //[│   └─TableFullScan_13 100000.00 100000 cop[tikv] table:t1 tikv_task:{time:35.1ms, loops:0} keep order:false N/A N/A]
> //[└─TableReader_12(Probe) 1.00 1 root  time:45.4µs, loops:2, cop_task: {num: 1, max: 116.4µs, proc_keys: 0, rpc_num: 1, rpc_time: 107.6µs, copr_cache_hit_ratio: 0.00} data:Selection_11 174 Bytes N/A]
> //[└─Selection_11 1.00 1 cop[tikv]  tikv_task:{time
> //
> 
> use inner to build
> //[HashJoin_9 0.80 0 root  time:55.5ms, loops:1, **build_hash_table:{total:198.1µs, fetch:190.9µs, build:7.17µs}**, probe:{concurrency:5, total:276.8ms, max:55.4ms, probe:17.8ms, fetch:258.9ms} semi join, equal:[eq(test.t2.name, test.t1.name)] 17.5 KB 0 Bytes]
> //[├─TableReader_12(Build) 1.00 1 root  time:126.1µs, loops:2, cop_task: {num: 1, max: 234.9µs, proc_keys: 0, rpc_num: 1, rpc_time: 214.8µs, copr_cache_hit_ratio: 0.00} data:Selection_11 174 Bytes N/A]
> //[│ └─Selection_11 1.00 1 cop[tikv]  tikv_task:{time:201.2µs, loops:0} not(isnull(test.t2.name)) N/A N/A]
> //[│   └─TableFullScan_10 1.00 1 cop[tikv] table:t2 tikv_task:{time:201.2µs, loops:0} keep order:false N/A N/A]
> //[└─TableReader_15(Probe) 100000.00 100000 root  time:53.8ms, loops:99, cop_task: {num: 1, max: 49.5ms, proc_keys: 0, tot_proc: 48ms, rpc_num: 1, rpc_time: 49.4ms, copr_cache_hit_ratio: 0.00} data:Selection_14 2.58 MB N/A]
> //[└─Selection_14 100000.00 100000 cop[tikv]  tikv_task:{time:
> ```

Yes, there's already an implementation in TiDB to support building the hash table on the inner side of the INNER JOIN.

This time we want to support it for (anti) (outer) semi join. It's a very similar job. But may need to have some changes to make it satisfy the property of the semi join.

**Comment by winoros** [comment]
> Hi, can i have a try this issue? And what we should do is add extra plan like `joins = append(joins, p.getHashJoin(prop, 1, true))`? Because i ran below test, and enforce to use inner or outer to build hash table(make `getHashJoin` third arg true or false).
> 
> ```
> func TestJoin(t *testing.T) {
> 	mock, err := mock.NewCluster()
> 	require.NoError(t, err)
> 	require.NoError(t, mock.Start())
> 	defer mock.Stop()
> 
> 	tk := testkit.NewTestKit(t, mock.Storage)
> 	tk.MustExec("use test")
> 	tk.MustExec("create table t1(id int,name char(128));")
> 	tk.MustExec("create table t2(name char(128));")
> 	for i := 0; i < 100000; i++ {
> 		id := strconv.FormatInt(int64(i), 10)
> 		tk.MustExec("insert into t1(id,name) values(" + id + "," +
> 			"'" + strings.Repeat("xx", 10) + id + "');")
> 	}
> 	tk.MustExec("insert into t2(name) values('name1')")
> 	tk.MustExec("analyze table t1")
> 	tk.MustExec("analyze table t2")
> 	tk.MustQuery("explain analyze SELECT * FROM t2 where exists (SELECT t2.name FROM t1 where t2.name = t1.name)").Check(nil)
> }
> ```
> 

yes, the changes in the planner side is rather small, just add one more possible physical plan.

**Comment by likzn** [comment]
@winoros Hi, i link a PR with this issue. PTAL if free.

**Comment by likzn** [comment]
@winoros @fixdb @XuHuaiyu Hi, Can you help watching with this PR. i think it's been a while. Thanks

**Comment by fixdb** [comment]
Hi @likzn, thanks for your patience. We will get it very soon.

**Comment by likzn** [comment]
@winoros @fixdb @XuHuaiyu Hi, PTAL.
